File: algorithms/c#/MergeSort.cs
namespace Algorithms
{
	public static class MergeSort
	{
		private static void Merge(int[] array, int start, int median, int end)
		{
			int index = 0;
			int left = start;
			int right = median + 1;
			int size = (end - start) + 1;

			int[] segment = new int[size];

			while(left <= median && right <= end)
			{
				if(array[left] <= array[right]) segment[index++] = array[left++];
				else segment[index++] = array[right++];
			}

			while(left <= median)
			{
				segment[index++] = array[left++];
			}

			while(right <= end)
			{
				segment[index++] = array[right++];
			}

			for(int i = 0; i < size; i++)
			{
				array[start + i] = segment[i];
			}
		}

		public static void Sort(int[] array, int start, int end)
		{
			if(start < end)
			{
				int median = (start + end) / 2;

				Sort(array, start, median);
				Sort(array, median + 1, end);
				Merge(array, start, median, end);
			}
		}
	}
}

File: algorithms/go/merge_sort.go
package algorithms

func Merge(slice []int, start int, median int, end int) {
	index := 0
	left := start
	right := median + 1
	size := (end - start) + 1

	segment := make([]int, size)

	for left <= median && right <= end {
		if slice[left] <= slice[right] {
			segment[index] = slice[left]
			left++
			index++
		} else {
			segment[index] = slice[right]
			right++
			index++
		}
	}

	for left <= median {
		segment[index] = slice[left]
		left++
		index++
	}

	for right <= end {
		segment[index] = slice[right]
		right++
		index++
	}

	for i := 0; i < size; i++ {
		slice[start + i] = segment[i]
	}
}

func MergeSort(slice []int, start int, end int) {
	if start < end {
		median := (start + end) / 2

		MergeSort(slice, start, median)
		MergeSort(slice, median + 1, end)
		Merge(slice, start, median, end)
	}
}